compiler/rustc_codegen_ssa/src/errors.rs

```diff
@@ -0,0 +1,197 @@
+}
+
+// Reports Paths using `Debug` implementation rather than Path's `Display` implementation.
+#[derive(Diagnostic)]
+#[diag(codegen_ssa::copy_path)]
+pub struct CopyPath<'a> {
+    from: DebugArgPath<'a>,
+    to: DebugArgPath<'a>,
+    error: Error,
+}
+
+impl<'a> CopyPath<'a> {
+    pub fn new(from: &'a Path, to: &'a Path, error: Error) -> CopyPath<'a> {
+        CopyPath { from: DebugArgPath(from), to: DebugArgPath(to), error }
+    }
+}
+
+struct DebugArgPath<'a>(pub &'a Path);
+
+impl IntoDiagnosticArg for DebugArgPath<'_> {
+    fn into_diagnostic_arg(self) -> rustc_errors::DiagnosticArgValue<'static> {
+        DiagnosticArgValue::Str(Cow::Owned(format!("{:?}", self.0)))
+    }
+}
+
+#[derive(Diagnostic)]
+#[diag(codegen_ssa::ignoring_emit_path)]
+pub struct IgnoringEmitPath {
+    pub extension: String,
+}
+
+#[derive(Diagnostic)]
+#[diag(codegen_ssa::ignoring_output)]
+pub struct IgnoringOutput {
+    pub extension: String,
+}
+
+#[derive(Diagnostic)]
+#[diag(codegen_ssa::create_temp_dir)]
+pub struct CreateTempDir {
+    pub error: Error,
+}
+
+#[derive(Diagnostic)]
+#[diag(codegen_ssa::incompatible_linking_modifiers)]
+pub struct IncompatibleLinkingModifiers;
+
+#[derive(Diagnostic)]
+#[diag(codegen_ssa::add_native_library)]
+pub struct AddNativeLibrary<'a> {
+    pub library_path: &'a str,
+    pub error: Error,
+}
+
+#[derive(Diagnostic)]
+#[diag(codegen_ssa::multiple_external_func_decl)]
+pub struct MultipleExternalFuncDecl<'a> {
+    #[primary_span]
+    pub span: Span,
+    pub function: Symbol,
+    pub library_name: &'a str,
+}
+
+pub enum LinkRlibError {
+    MissingFormat,
+    OnlyRmetaFound { crate_name: Symbol },
+    NotFound { crate_name: Symbol },
+}
+
+impl IntoDiagnostic<'_, !> for LinkRlibError {
+    fn into_diagnostic(self, handler: &Handler) -> DiagnosticBuilder<'_, !> {
+        match self {
+            LinkRlibError::MissingFormat => {
+                handler.struct_fatal(fluent::codegen_ssa::rlib_missing_format)
+            }
+            LinkRlibError::OnlyRmetaFound { crate_name } => {
+                let mut diag = handler.struct_fatal(fluent::codegen_ssa::rlib_only_rmeta_found);
+                diag.set_arg("crate_name", crate_name);
+                diag
+            }
+            LinkRlibError::NotFound { crate_name } => {
+                let mut diag = handler.struct_fatal(fluent::codegen_ssa::rlib_not_found);
+                diag.set_arg("crate_name", crate_name);
+                diag
+            }
+        }
+    }
+}
+
+#[derive(Diagnostic)]
+#[diag(codegen_ssa::thorin_dwarf_linking)]
+#[note]
+pub struct ThorinDwarfLinking {
+    pub thorin_error: ThorinErrorWrapper,
+}
+pub struct ThorinErrorWrapper(pub thorin::Error);
+
+// FIXME: How should we support translations for external crate errors?
+impl IntoDiagnosticArg for ThorinErrorWrapper {
+    fn into_diagnostic_arg(self) -> DiagnosticArgValue<'static> {
+        DiagnosticArgValue::Str(Cow::Owned(format!("{:?}", self.0)))
+    }
+}
```

Comment:
This `thorin::Error` diagnostic is currently untranslatable. [thorin::Error already implements Display](https://github.com/rust-lang/thorin/blob/c454f85084a882c3905d838b0178b5e2953fe08e/thorin/src/error.rs#L167), which brings the question:

### How do we want to support translatable diagnostics when using an external crate ?

I think the correct path could be for Errors to be able to provide their own translated message, now, this probably involves additions to the standard library and a long-term process for this practice to be adopted in the ecosystem.

Another solution could be for `rustc` to handle this internally for the crate it uses, for example, we could match over the error cases and provide our own keys and translations, but this is sub-optional to say the least. 
`rustc` should not be maintaining translations for other crates, and at the end this is something that is going to impact the ecosystem once the compiler creates the expectation of having i18n interfaces.

What do you folks think?

Reply:
`thorin` is a crate that I wrote and that lives in `rust-lang/thorin` so we can make whatever changes we need to it - in general, I'm hoping that we can pattern match on error types and just map those to whatever Fluent messages we want.

Reply:
Sounds ok, but we risk getting diagnostics out of sync between the two repositories

Reply:
Do we want to include all of `thorin::Error` keys `codegen_ssa.ftl` file, or do you think it would be better to have `thorin.ftl` and a separate `thorin_error.rs` file inside `codegen_ssa` ?

I prefer the latter, but I think our "rule" so far has been one `.ftl` file per crate.

Reply:
This is probably a naive question, but is it possible to put the `.ftl` file in the thorin crate?

Reply:
I guess we could add a constant to Thorin with the ftl content but I don't think we'd gain too much from doing that, because we'd then want to run it through the `fluent_messages!` macro to generate our typed identifiers and stuff like that (which would require some changes). For now, we can just keep them in `codegen_ssa.ftl`, `thorin` is unlikely to many new errors or have its messages change. There's a one-off task in #100717 for keeping each ftl file with its crate rather than in `rustc_error_messages` - whatever we do for that would likely resolve this too.

Reply:
Addressed [here](https://github.com/rust-lang/rust/pull/102612/commits/f69f1b0fc8f8d730798b9633f0f1114d96259bf4)